src/pb/publisher.rs

```diff
@@ -2,45 +2,131 @@
+    instance: &'a str,
+    event: &'a str,
+    options: PublisherOptions,
+    phantom: PhantomData<T>,
+}
+
+impl<'a, T: POD> PublisherBuilder<'a, T> {
+    pub fn new(service: &'a str, instance: &'a str, event: &'a str) -> Self {
+        Self {
+            service,
+            instance,
+            event,
+            options: PublisherOptions::default(),
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn history_capacity(mut self, history_capacity: u64) -> Self {
+        self.options.history_capacity = history_capacity;
+        self
+    }
+
+    pub fn node_name(mut self, node_name: String) -> Self {
+        self.options.node_name = node_name;
+        self
+    }
+
+    pub fn subscriber_too_slow_policy(
+        mut self,
+        subscriber_too_slow_policy: ConsumerTooSlowPolicy,
+    ) -> Self {
+        self.options.subscriber_too_slow_policy = subscriber_too_slow_policy;
+        self
+    }
+
+    pub fn create(mut self) -> Result<Publisher<T>, IceoryxError> {
+        self.options.offer_on_create = true;
+        let ffi_pub = FfiPublisher::new(self.service, self.instance, self.event, &self.options)
+            .ok_or(IceoryxError::PublisherCreationFailed)?;
+
+        Ok(Publisher {
+            ffi_pub,
+            phantom: PhantomData,
+        })
+    }
+
+    pub fn create_without_offer(mut self) -> Result<InactivePublisher<T>, IceoryxError> {
+        self.options.offer_on_create = false;
+        let ffi_pub = FfiPublisher::new(self.service, self.instance, self.event, &self.options)
+            .ok_or(IceoryxError::PublisherCreationFailed)?;
+
+        Ok(InactivePublisher {
+            ffi_pub,
+            phantom: PhantomData,
+        })
+    }
+}
+
+pub struct InactivePublisher<T: POD> {
+    ffi_pub: Box<FfiPublisher>,
+    phantom: PhantomData<T>,
+}
+
+impl<T: POD> InactivePublisher<T> {
+    fn new_from_publisher(publisher: Publisher<T>) -> Self {
+        Self {
+            ffi_pub: publisher.ffi_pub,
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn offer(self) -> Publisher<T> {
+        self.ffi_pub.offer();
+        Publisher::new_from_inactive_publisher(self)
+    }
+}
 
 pub struct Publisher<T: POD> {
-    publisher: Topic<T>,
+    ffi_pub: Box<FfiPublisher>,
+    phantom: PhantomData<T>,
 }
 
 impl<T: POD> Publisher<T> {
-    pub(super) fn new(publisher: Topic<T>) -> Self {
-        Publisher { publisher }
+    fn new_from_inactive_publisher(publisher: InactivePublisher<T>) -> Self {
+        Self {
+            ffi_pub: publisher.ffi_pub,
+            phantom: PhantomData,
+        }
     }
 
     pub fn is_offered(&self) -> bool {
-        self.publisher.ffi_pub.is_offered()
+        self.ffi_pub.is_offered()
     }
 
-    pub fn stop(self) -> Topic<T> {
-        self.publisher.ffi_pub.stop_offer();
-        self.publisher
+    pub fn stop(self) -> InactivePublisher<T> {
+        self.ffi_pub.stop_offer();
+        InactivePublisher::new_from_publisher(self)
```

Comment:
UUhh this is nice, actually we could also implement such a thing for c++.

```cpp
struct Publisher {
  static InactivePublisher stop(Publisher&& p);
};
```

Reply:
This is the type state pattern. Unfortunately not fully possible with C++ since it is not fully possible to control the ownership.

The way this works in Rust is that the compiler will complain if you use the publisher after calling `stop_offer`